### ISSUE TITLE
Inconsistent user sorting in account users, course users, and course gradebook

### ISSUE THREAD

**Issue by merichar** [instructure/canvas-lms#1229]
### Summary:

Instructure-hosted instances of Canvas do not sort hyphenated surnames correctly. e.g. Alice Smith-Nachos is placed before Alice Smith. When using the open source version backed by MySQL or PostgreSQL, hyphenated surnames are returned in a different (APA-compliant) order. Given that ActiveRecord's order method is used for User.order_by_sortable_name, my suspicion is that the issue lies in trusting all databases to sort edge cases consistently.

https://github.com/instructure/canvas-lms/blob/c63a4c51ab141824ea83343d8f335842aa40b738/app/models/user.rb#L250

#### Screenshots of hyphenated surnames in a hosted instances:

User search under account
![canvas-alpha-account-people](https://user-images.githubusercontent.com/304241/35905895-c161b1d4-0bb6-11e8-8e30-f3fd1f75eaca.png)

Gradebook under course
![canvas-alpha-course-gradebook](https://user-images.githubusercontent.com/304241/35905897-c177ee7c-0bb6-11e8-8390-745b1587e7a1.png)

People under course
![canvas-alpha-course-people](https://user-images.githubusercontent.com/304241/35905898-c18642b0-0bb6-11e8-9d09-b663194abe9d.png)

### Steps to reproduce:

 0. Add "Alice Smith" to a hosted Canvas instance
 1. Add "Alice Smith-Nachos" to the same instance
 2. Go to the People link under the account where the users are
 3. Notice the ordering of names

### Expected behavior:

The APA (and Chicago Manual of Style and the MLA handbook) style for all examples in the screenshots above would place Lorelei Hoover before Joseph Hoover-Mertz. If the intent was to go against the style handbooks, the sorting would still be inconsistent between the hosted version and recommended configuration of open source version of Canvas.

### Additional notes:

APA style discussion for ordering hyphenated surnames: 
http://blog.apastyle.org/apastyle/2010/05/alphabetization-in-apa-style.html

**Comment by ccutrer** [comment]
Hosted Canvas uses the Unicode Collation Algorithm for sorting http://unicode.org/reports/tr10/ in all locations. If you find two different places in the UI that sort differently, that needs to be addressed as a bug. However, insisting on an arbitrary style manual outside of Unicode is not something we are currently considering - it would compromise sorting of non-English names (when used in English, or if the we're using language specific sorting rules because the user is using a different language for Canvas).

For self-hosted, we attempt to use the same sorting when possible, but it's not always possible, as it depends on additional support from the browser, the operating system the server is running, and additional extensions available on the database server. In particular, [pg_collkey](http://www.public-software-group.org/pg_collkey) and the [ICU](http://site.icu-project.org) library must be available.

**Comment by merichar** [comment]
I will follow-up through our rep, but to insist that the unicode sorting of hyphenated surnames is the right thing to do, instead of respecting a user's locale and language norms, is mind-boggling. It is not ethnocentric to expect that English names be sorted using English style, and to call MLA, APA, and Chicago style arbitrary in this context is intentionally obtuse. Would you have used the same argument if through some happenstance unicode sorting had placed 'b' before 'a'? We are not taught to sort names using the unicode algorithm in US schools, and users expect to find a sorted name in the culturally-established order.

Consider Ruby's own default sorting of unicode strings:
`2.4.1 :010 > ["\u0061","\u0062","\u0061\u002d\u0062"].sort`
` => ["a", "a-b", "b"]`
If the language backing the web framework you use has conceded the reasonableness of this ordering, subjecting an unexpected ordering on a significant portion of your user base should be addressed, at the very least, less flippantly than you have done above.

**Comment by ccutrer** [comment]
I apologize for taking so long to respond here - I had missed the notification e-mail.

I think you're misunderstanding what Unicode collation is. It is, by very definition, "respecting a user's locale and language norms". That's the whole reason it exists. Canvas does _not_ use Ruby's own default sorting of unicode strings as you say because it is a byte-string sorting, and not a sorting according to locale rules. The example you used only happens to work for your desired purpose because in ASCII a hyphen just happens to have a lower code point than latin characters. But this is the exact same reason that in raw ruby, all capital letters sort before all lower case letters, and a major reason we don't use byte-string sorting:

```
irb(main):001:0> ["Mcgraw, Abe", "McGraw, Hill"].sort
=> ["McGraw, Hill", "Mcgraw, Abe"]
```

Would you not be up in arms that capitalization should be ignored for sorting purposes, and thus `Mcgraw, Abe` should sort before `McGraw, Hill`? Byte-string sorting gets even worse once you start incorporating accented characters. For example, in English, we would sort `ñ` as if there is no accent at all (just move on to the next character to see if there is a difference, mixing `n` and `ñ` together in long lists). In Spanish though, they consider it a separate character, that sorts immediately after `n`. All `n`s are grouped together, and then all `ñ`s. However, a raw byte-string sort will sort it wrong in both languages - as a separate character yes, but also a separate character _after_ `z`, since it doesn't even fit in the code point sequence between `a` and `z`

The Unicode collation algorithm fixes both of these problems. And many more. It's far more complex than simply re-arranging the order of characters. It can say things like "sort `ñ` the same as `n`, except if the rest of the string is identical, then group `ñ` after `n`". Or "consider lower case and upper case the same. Oh and here's what that even means, in the context of every language we cover". Or "ignore punctuation when sorting. Unless everything else is equal, then here is how you sort punctuation". And the relative ordering of all those rules.  If you want to get into the nitty gritty details, they're at http://www.unicode.org/reports/tr10/. But basically it boils down to a bunch of rules (which different languages have different default sets of rules), and several higher level settings of how to process those rules slightly differently (see http://icu-project.org/apiref/icu4c/ucol_8h.html#a583fbe7fc4a850e2fcc692e766d2826c, which in Canvas corresponds to https://github.com/instructure/canvas-lms/blob/stable/lib/canvas/icu.rb#L75). For example, in English the rules are `` (look at that, English has no special cases!). For Spanish, the rules are `&N<ñ<<<Ñ` (I don't actually know what each character there means, but it's changing the sorting rules for `ñ`).

Finally, we apply some processing settings:
```
        # Reference documentation (some option names differ in ruby-space)for these options is at
        # http://userguide.icu-project.org/collation/customization#TOC-Default-Options
        collator.normalization_mode = true
        collator.numeric_collation = true
        collator.strength = :tertiary
        collator
```

We have a ticket open internally to look at those settings, and see if they're negatively impacting the situation you describe (hyphenated names). But it's something we don't touch lightly, because it also affects the comma that is used for separating surnames before given names. I.e. you want `Hoover, Lorelei` to sort prior to `Hoover-Mertz, Joseph`, but you _don't_ want `Hoover-Lorelei, Joseph` to sort prior to `Hoover, Lorelei Mary` (i.e. ignoring all punctuation would end up with Mary sorting after Joseph).

In summary, I'm sorry if you feel sorting is being taken flippantly. I assure you this is the absolute opposite of the truth. A lot of thought and effort has gone into sorting properly--both generally, and language specific anachronisms. And we build on a lot of work done by the Unicode Consortium. And due to all that work, we're cautious of "flippant" user requests that at face value look like someone seeing something they don't like, and wanting things to change without understanding all the reasoning for how things are.
